packages/vite/src/node/build.ts

```diff
@@ -875,13 +889,66 @@ async function buildEnvironment(
       prepareOutDir(resolvedOutDirs, emptyOutDir, environment)
     }
 
-    const res: RolldownOutput[] = []
-    for (const output of normalizedOutputs) {
-      res.push(await bundle[options.write ? 'write' : 'generate'](output))
-    }
+    const res = await build()
+
     logger.info(
       `${colors.green(`✓ built in ${displayTime(Date.now() - startTime)}`)}`,
     )
+
+    async function build() {
+      const res: RolldownOutput[] = []
+      for (const output of normalizedOutputs) {
+        // TODO(underfin): using the generate at development build could be improve performance.
+        res.push(await bundle![options.write ? 'write' : 'generate'](output))
+      }
+
+      if (server) {
+        // watching the files
+        for (const file of bundle!.watchFiles) {
+          if (path.isAbsolute(file) && fs.existsSync(file)) {
+            server.watcher.add(file)
+          }
+        }
+
+        // Write the output files to memory
+        for (const output of res) {
+          for (const outputFile of output.output) {
+            server.memoryFiles[outputFile.fileName] =
+              outputFile.type === 'chunk' ? outputFile.code : outputFile.source
+          }
+        }
+      }
+      return res
+    }
+
+    if (server) {
+      server.watcher.on('change', async (file) => {
+        const startTime = Date.now()
+        const hmrOutput = (await bundle!.generateHmrPatch([file]))!
+        if (hmrOutput.patch) {
+          const url = `${Date.now()}.js`
+          server.memoryFiles[url] = hmrOutput.patch
+          server.ws.send({
+            type: 'update',
+            updates: hmrOutput.hmrBoundaries.map((boundary) => {
+              return {
+                type: 'js-update',
+                url,
+                path: boundary.boundary,
+                acceptedPath: boundary.acceptedVia,
+                timestamp: 0,
+              }
+            }),
+          })
+          logger.info(
+            `${colors.green(`✓ Found ${path.relative(root, file)} changed, rebuilt in ${displayTime(Date.now() - startTime)}`)}`,
+          )
+
+          await build()
+        }
+      })
```

Comment:
If multiple files were changed almost the same time (e.g. `git checkout`), the handler would run concurrently. That would probably cause a problem. For example, `server.ws.send` to be called in the reverse order.

Reply:
The rolldown side has a lock at https://github.com/rolldown/rolldown/blob/main/crates/rolldown_binding/src/bundler.rs#L121. So it only one file could be into `hmr/builld` process.